File: python/src/day05.py
#!/usr/bin/env python

from pathlib import Path
from typing import List

INPUT_FILE = Path.cwd().parent / "inputs" / "day05.txt"


def transform_input(input: str) -> List[str]:
    return input.strip().split("\n")


def part1(passes: List[str]) -> int:
    return max(transform_to_costs(passes))


def transform_to_costs(passes: List[str]) -> List[int]:
    return [calculate_cost(x) for x in passes]


def calculate_cost(x: str) -> int:
    bin_ = x.replace("F", "0").replace("B", "1").replace("L", "0").replace("R", "1")
    return int(bin_, 2)


def part2(passes: List[str]) -> int:
    costs = set(transform_to_costs(passes))
    for x in range(min(costs) + 1, max(costs) - 1):
        if x not in costs and x - 1 in costs and x + 1 in costs:
            return x

    raise Exception("Impossible!")


if __name__ == "__main__":
    with INPUT_FILE.open("r") as f:
        input = transform_input(f.read())

    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

File: rust/src/day05.rs
use itertools::Itertools;

static INPUT: &str = include_str!("../../inputs/day05.txt");

type SeatID = u32;

pub fn run() {
    let mut seat_ids = transform_input(INPUT);

    println!("Part 1: {}", part1(&seat_ids));
    println!("Part 2: {}", part2(&mut seat_ids));
}

/// Convert each line in the input into a seat ID. We replace each letter with its
/// corresponding bit, and from that we calculate the seat ID.
///
fn transform_input(input: &str) -> Vec<SeatID> {
    input
        .trim_end()
        .split('\n')
        .map(|x| {
            let bin = x
                .replace(&['F', 'L'][..], "0")
                .replace(&['B', 'R'][..], "1");

            u32::from_str_radix(&bin, 2).unwrap()
        })
        .collect()
}

fn part1(seat_ids: &[SeatID]) -> SeatID {
    *seat_ids.iter().max().unwrap()
}

fn part2(seat_ids: &mut [SeatID]) -> SeatID {
    seat_ids.sort_unstable();

    for (id, next_id) in seat_ids.iter().tuple_windows() {
        if id + 2 == *next_id {
            return id + 1;
        }
    }

    panic!("(Part 2) No seat ID found.");
}
